File: lib/accounting/engine/profiles.ts
import type { BankCapability, BankProfileId, ParserCapabilityMatrix } from "@/lib/accounting/engine/types";

export type BankParserProfile = {
  id: BankProfileId;
  bankName: string;
  statementType: string;
  parserVersion: string;
  capabilities: ParserCapabilityMatrix;
};

const enabledCapabilities = (items: BankCapability[]): ParserCapabilityMatrix => {
  const set = new Set(items);
  return {
    ocr_required: set.has("ocr_required"),
    supports_multi_page: set.has("supports_multi_page"),
    supports_combined_statements: set.has("supports_combined_statements"),
    running_balance_validation: set.has("running_balance_validation"),
    vat_extraction: set.has("vat_extraction"),
    ai_categorisation: set.has("ai_categorisation"),
    review_mode: set.has("review_mode"),
    bank_charges_detection: set.has("bank_charges_detection"),
  };
};

export const BANK_PARSER_PROFILES: BankParserProfile[] = [
  {
    id: "fnb_business_v1",
    bankName: "FNB South Africa",
    statementType: "business_bank_statement",
    parserVersion: "fnb_business_v1",
    capabilities: enabledCapabilities([
      "ocr_required",
      "supports_multi_page",
      "supports_combined_statements",
      "running_balance_validation",
      "vat_extraction",
      "ai_categorisation",
      "review_mode",
      "bank_charges_detection",
    ]),
  },
  {
    id: "standard_bank_business_v1",
    bankName: "Standard Bank",
    statementType: "business_bank_statement",
    parserVersion: "standard_bank_business_v1",
    capabilities: enabledCapabilities(["ocr_required", "supports_multi_page", "review_mode"]),
  },
  {
    id: "absa_business_v1",
    bankName: "ABSA",
    statementType: "business_bank_statement",
    parserVersion: "absa_business_v1",
    capabilities: enabledCapabilities(["ocr_required", "supports_multi_page", "review_mode"]),
  },
  {
    id: "nedbank_business_v1",
    bankName: "Nedbank",
    statementType: "business_bank_statement",
    parserVersion: "nedbank_business_v1",
    capabilities: enabledCapabilities(["ocr_required", "supports_multi_page", "review_mode"]),
  },
  {
    id: "capitec_business_v1",
    bankName: "Capitec",
    statementType: "business_bank_statement",
    parserVersion: "capitec_business_v1",
    capabilities: enabledCapabilities(["ocr_required", "supports_multi_page", "review_mode"]),
  },
  {
    id: "investec_business_v1",
    bankName: "Investec",
    statementType: "business_bank_statement",
    parserVersion: "investec_business_v1",
    capabilities: enabledCapabilities(["ocr_required", "supports_multi_page", "review_mode"]),
  },
];

export function getParserProfileById(id: BankProfileId) {
  return BANK_PARSER_PROFILES.find((profile) => profile.id === id) ?? null;
}

File: workers/accounting_worker/engine/profiles.py
from __future__ import annotations

from dataclasses import dataclass

from .registry import BankParser, ParserCapabilities, ParserProfile


@dataclass(frozen=True)
class KeywordBankParser(BankParser):
    profile: ParserProfile
    keywords: tuple[str, ...]

    def matches(self, text_sample: str, file_name: str) -> bool:
        haystack = f"{text_sample} {file_name}"
        return any(keyword in haystack for keyword in self.keywords)


FNB_PARSER = KeywordBankParser(
    profile=ParserProfile(
        id="fnb_business_v1",
        bank_name="FNB South Africa",
        version="fnb_business_v1",
        statement_type="business_bank_statement",
        capabilities=ParserCapabilities(
            ocr_required=True,
            supports_multi_page=True,
            supports_combined_statements=True,
            running_balance_validation=True,
            vat_extraction=True,
            ai_categorisation=True,
            review_mode=True,
            bank_charges_detection=True,
        ),
    ),
    keywords=("fnb", "first national bank", "platinum business account"),
)

PARSER_SCAFFOLDS = [
    KeywordBankParser(
        profile=ParserProfile(
            id="standard_bank_business_v1",
            bank_name="Standard Bank",
            version="standard_bank_business_v1",
            statement_type="business_bank_statement",
            capabilities=ParserCapabilities(
                ocr_required=True,
                supports_multi_page=True,
                supports_combined_statements=False,
                running_balance_validation=False,
                vat_extraction=False,
                ai_categorisation=False,
                review_mode=True,
                bank_charges_detection=False,
            ),
        ),
        keywords=("standard bank",),
    ),
    KeywordBankParser(
        profile=ParserProfile(
            id="absa_business_v1",
            bank_name="ABSA",
            version="absa_business_v1",
            statement_type="business_bank_statement",
            capabilities=ParserCapabilities(
                ocr_required=True,
                supports_multi_page=True,
                supports_combined_statements=False,
                running_balance_validation=False,
                vat_extraction=False,
                ai_categorisation=False,
                review_mode=True,
                bank_charges_detection=False,
            ),
        ),
        keywords=("absa",),
    ),
    KeywordBankParser(
        profile=ParserProfile(
            id="nedbank_business_v1",
            bank_name="Nedbank",
            version="nedbank_business_v1",
            statement_type="business_bank_statement",
            capabilities=ParserCapabilities(
                ocr_required=True,
                supports_multi_page=True,
                supports_combined_statements=False,
                running_balance_validation=False,
                vat_extraction=False,
                ai_categorisation=False,
                review_mode=True,
                bank_charges_detection=False,
            ),
        ),
        keywords=("nedbank",),
    ),
    KeywordBankParser(
        profile=ParserProfile(
            id="capitec_business_v1",
            bank_name="Capitec",
            version="capitec_business_v1",
            statement_type="business_bank_statement",
            capabilities=ParserCapabilities(
                ocr_required=True,
                supports_multi_page=True,
                supports_combined_statements=False,
                running_balance_validation=False,
                vat_extraction=False,
                ai_categorisation=False,
                review_mode=True,
                bank_charges_detection=False,
            ),
        ),
        keywords=("capitec",),
    ),
    KeywordBankParser(
        profile=ParserProfile(
            id="investec_business_v1",
            bank_name="Investec",
            version="investec_business_v1",
            statement_type="business_bank_statement",
            capabilities=ParserCapabilities(
                ocr_required=True,
                supports_multi_page=True,
                supports_combined_statements=False,
                running_balance_validation=False,
                vat_extraction=False,
                ai_categorisation=False,
                review_mode=True,
                bank_charges_detection=False,
            ),
        ),
        keywords=("investec",),
    ),
]
